### ISSUE TITLE
Expose code monitor, saved search, extend and other functions to non-authenticated users and add signup CTA

### ISSUE THREAD

**Issue by rrhyne** [sourcegraph/sourcegraph#21093]
These features represent reasons to sign up, and we should use them as an opportunity to show unauthenticated users portions of the product they are missing out on.

Add code monitor and save search buttons to UI, but when moused over, display a popup stating: 'Signup to use this feature' or similar. 

- [Figma](https://www.figma.com/file/LfQUGTLvfYWxyuFoyQsptB/Expose-features?node-id=0%3A1&viewport=-81%2C-640%2C0.5)
- Metrics for cloud are defined in the Figma file, we use the Flow_Stage_NameOfStage convention

We'll A/B test under the `w0-signup-optimisation` flag.

There are two parts for this that can be done in parallel:

- [x] Expose auth-only functionality to unauthenticated users in search results and redirect to new sign up page (owned by @novoselrok) #22091
  - [x] Show "Monitor" to unauthenticated users
     - For unauthenticated users only, the button will always be enabled and show the CTA, regardless of whether the search allows creating monitors or not.
  - [x] Show "Save search" to unauthenticated users
    - This also included modifying the "Saved search" icon to a bookmark (for all users with and without the feature flag) 
  - [x] Add an "Extend" button visible to all users; when clicked by authenticated users it goes to the `/extensions` page
  - [x] Add a sign-up CTA when these buttons are clicked by unauthenticated users 
  - [x] Modify the CTA for search contexts for unauthenticated users to use the new sign up page
  - [x] Add a new CTA at the top of search results for unauthenticated users to create an account to add their own repos
    - This CTA would be shown on a cadence such as: displayed for every search on the first 5 or 10 searches, then once every 5 searches.
  - [x] Add a link under the "search snippets" section of the sidebar prompting users to create an account to edit their snippets. This redirects to the standard sign up page. (Lower priority)
- [x] Create a new sign up page when the CTA is clicked  (owned by @limitedmage)
  - [x] Basic, non-functional page #22104
  - [x] Users can create new account or sign up via GitHub & GitLab in the page #22135
  - [x] The text content in the page is different based on which CTA was clicked #22135
  - [x] Events are logged when users create an account, either from GitHub/GitLab oauth or with email&password #22135
  - [x] Once users sign up, they should be redirected back to the functionality they were intending to use (create a monitor or a saved search), or at least to the search they were on previously
    - The "Monitor" button should redirect to [code monitors](https://sourcegraph.com/code-monitoring)
    - The "Save search" button should redirect to [code monitors](https://sourcegraph.com/code-monitoring)
    - The "Extend" button should redirect to [extensions](https://sourcegraph.com/extensions?category=All)
    - Public code CTA (both in the search contexts menu and at the top of the search results) takes users to [repositories page](https://sourcegraph.com/users/rrhyne/settings/repositories)

**Comment by malomarrec** [comment]
I this lands, we'll A/B test under the `w0-signup-optimisation` flag

**Comment by limitedmage** [comment]
@rrhyne  "Extend" doesn't exist currently so it should not be part of this work.

**Comment by novoselrok** [comment]
> The "Save search" button should redirect to code monitors

We could also redirect back to the search where user clicked on the "Save search" button.

**Comment by limitedmage** [comment]
> We could also redirect back to the search where user clicked on the "Save search" button.

I discussed it with @rrhyne  and I like this better actually.
